File: OpenGUI-UnityScript/Scripts/OGPopUp.js
#pragma strict

@script AddComponentMenu ("OpenGUI/PopUp")

class OGPopUp extends OGWidget {	
	var title : String;
	var isUp = false;
	var options : String[];
	var messageTarget : GameObject;
	var message : String;
	var passSelectedOption : boolean = false;
	var padding : Vector2 = new Vector2 ( 8.0, 8.0 );
	var indicator : String = "";
	
	var selectedOption : String;
	@HideInInspector var originalZ : float = 99;
	
	// Set options
	public function SetOptions ( strings : String[] ) {
		yield WaitForEndOfFrame();
		
		options = strings;
	}
				
	// Draw
	override function Draw ( x : float, y : float ) {	
		if ( !guiStyle ) { guiStyle = GUI.skin.box; }
		
		var textStyle : GUIStyle = new GUIStyle();
		textStyle.font = GUI.skin.label.font;
		textStyle.normal.textColor = guiStyle.normal.textColor;
		textStyle.fontSize = guiStyle.fontSize;
		textStyle.alignment = guiStyle.alignment;
		
		var activeStyle : GUIStyle = new GUIStyle ();
		activeStyle.normal.background = guiStyle.active.background;
		activeStyle.border = guiStyle.border;
		activeStyle.margin = guiStyle.margin;
		activeStyle.padding = guiStyle.padding;
																				
		if ( !isUp ) {			
			var label : String;
			
			if ( selectedOption ) {
				label = selectedOption;
			} else {
				label = title;
			}
			
			GUI.Box ( Rect ( x, y, transform.localScale.x + (guiStyle.padding.left * 2), transform.localScale.y ), "", guiStyle );
			
			GUI.Label ( Rect ( x + transform.localScale.x - 20, y + transform.localScale.y - 18, 20, 20 ), indicator, textStyle );
			
			if ( GUI.Button ( Rect ( x, y, transform.localScale.x + (guiStyle.padding.left * 2), transform.localScale.y ), label, textStyle ) ) {
				isUp = true;
			}
		} else {			
			GUI.Box ( Rect ( x, y, transform.localScale.x + (guiStyle.padding.left * 2), ( options.Length * (guiStyle.fontSize + guiStyle.padding.top) ) + ( guiStyle.padding.top * 2 ) ), "", activeStyle );
			
			for ( var i = 0; i < options.Length; i++ ) {			
				if ( GUI.Button ( Rect ( x + guiStyle.padding.left, y + guiStyle.padding.top + ( ( guiStyle.fontSize + guiStyle.padding.top ) * i ), transform.localScale.x + ( guiStyle.padding.left * 2 ), guiStyle.fontSize + guiStyle.padding.top ), options[i], textStyle ) ) {
					selectedOption = options[i];
					isUp = false;
					
					if ( messageTarget && message ) {
						if ( passSelectedOption ) {
							messageTarget.SendMessage ( message, selectedOption );
						} else {
							messageTarget.SendMessage ( message );
						}
					}
				}
			}
		}
	}
	
	// Update
	override function UpdateWidget () {		
		if ( originalZ == 99 ) {
			originalZ = transform.localPosition.z;
		}
		
		/*if ( isUp ) {
			transform.localPosition.z = -10;
		} else {
			transform.localPosition.z = originalZ;
		}*/
		
		if ( Input.GetKeyDown ( KeyCode.Escape ) ) {
			isUp = false;
		}
	}
}

File: C#/Scripts/OGPopUp.cs
using UnityEngine;
using System.Collections;

[AddComponentMenu ("OpenGUI/PopUp")]
public class OGPopUp : OGWidget {

	public string title;
	public bool isUp = false;
	public string[] options;
	public string message;
	public GameObject messageTarget;
	public Vector2 padding = new Vector2 ( 8.0f, 8.0f );
	
	public string selectedOption;
	float originalZ = 99;
				
	// Draw
	public override void Draw (  float x ,   float y   ){	
		if ( guiStyle == null ) { guiStyle = GUI.skin.box; }
		
		GUIStyle textStyle = new GUIStyle();
		textStyle.font = GUI.skin.label.font;
		textStyle.normal.textColor = guiStyle.normal.textColor;
		textStyle.fontSize = guiStyle.fontSize;
		textStyle.alignment = guiStyle.alignment;
								
		if ( !isUp ) {			
			string label;
			
			if ( selectedOption != "" ) {
				label = selectedOption;
			} else {
				label = title;
			}
			
			GUI.Box ( new Rect( x, y, transform.localScale.x + (guiStyle.padding.left * 2), transform.localScale.y ), "", guiStyle );
			
			if ( GUI.Button ( new Rect( x, y, transform.localScale.x + (guiStyle.padding.left * 2), transform.localScale.y ), label, textStyle ) ) {
				isUp = true;
			}
		} else {			
			GUI.Box ( new Rect( x, y, transform.localScale.x + (guiStyle.padding.left * 2), ( options.Length * (guiStyle.fontSize + padding.y) ) + ( guiStyle.padding.top * 2 ) ), "", guiStyle );
			
			for ( int i = 0; i < options.Length; i++ ) {			
				if ( GUI.Button ( new Rect( x + guiStyle.padding.left, y + padding.y * 0.5f + ( ( guiStyle.fontSize + padding.y ) * i ), transform.localScale.x + ( guiStyle.padding.left * 2 ), guiStyle.fontSize + guiStyle.padding.top ), options[i], textStyle ) ) {
					selectedOption = options[i];
					isUp = false;
					
					if ( messageTarget && message != "" ) {
						messageTarget.SendMessage ( message, i );
					}
				}
			}
		}
	}

	public void SetSelection ( int id  ){
		selectedOption = options[id];
		//label = selectedOption;
	}
	
	// Update
	public override void UpdateWidget (){		
		if ( originalZ == 99 ) {
			originalZ = transform.localPosition.z;
		}
		
		if ( isUp ) {
			transform.localPosition += new Vector3(0f, 0f, -10f);
		} else {
			transform.localPosition = new Vector3(transform.localPosition.x, transform.localPosition.y, originalZ);
		}
		
		if ( Input.GetKeyDown ( KeyCode.Escape ) ) {
			isUp = false;
		}
	}
}
